### ISSUE TITLE
Stacktrace error from heartbeat.ex

### ISSUE THREAD

**Issue by beerlington** [cjbottaro/faktory_worker_ex#28]
Hi,

I've been seeing some frequent errors in our production server logs, and I'm not really sure if this is something to be concerned with. Is this an error you've come across before?

```
Elixir.ArgumentError: argument error
  Module "erlang", in :erlang.iolist_to_binary/1
  File "lib/poison/parser.ex", line 35, in Poison.Parser.parse/2
  File "lib/poison/parser.ex", line 51, in Poison.Parser.parse!/2
  File "lib/poison.ex", line 83, in Poison.decode!/2
  File "lib/faktory/protocol.ex", line 84, in Faktory.Protocol.beat/2
  File "src/poolboy.erl", line 76, in :poolboy.transaction/3
  File "lib/faktory/heartbeat.ex", line 30, in Faktory.Heartbeat.beat/2
  File "lib/faktory/heartbeat.ex", line 24, in Faktory.Heartbeat.handle_info/2
```

**Comment by cjbottaro** [comment]
I have not seen that before. What version of Faktory are you using?

The only thing I can think of is that I messed up the protocol somehow and since worker processes and the heartbeat process share a pool of connections, that some connection has extraneous characters left over on its socket. I think this can also happen if a process dies midway though speaking the protocol and the connection gets put back into the pool.

I'm currently doing a major rewrite to greatly simplify a lot of things, one of them being not having a connection pool for the worker and heartbeat processes; instead each process will get its own dedicated (linked) connection.

I'm hoping to have the new version ready by this weekend.

**Comment by beerlington** [comment]
We're using 0.9.2 Pro. I haven't noticed it in development, just in production only. I'll keep an eye out for the next version. Thanks!

**Comment by cjbottaro** [comment]
@beerlington I have a new branch `v0.6.0` on Github. There are major breaking changes around how configuration is done. Also, I'm not happy with how connections work, so still working on that.

But all tests pass. And I won't push anything to that branch unless all the test pass, so maybe try it out (in non-prod first)?

Please read about the changes though:
* https://github.com/cjbottaro/faktory_worker_ex/blob/v0.6.0/CHANGELOG.md
* https://github.com/cjbottaro/faktory_worker_ex/blob/v0.6.0/README.md
* https://github.com/cjbottaro/faktory_worker_ex/blob/v0.6.0/lib/faktory/client.ex (see moduledoc)
* https://github.com/cjbottaro/faktory_worker_ex/blob/v0.6.0/lib/faktory/worker.ex (see moduledoc)

The tl;dr about configuration is that:
* Configuring clients/workers is pretty much exactly like configuring Ecto repos now
* You must manually put the clients/workers into your supervision tree now

**Comment by beerlington** [comment]
@cjbottaro that sounds great! The config was definitely something we struggled with in the beginning so I'll let you know how that update goes. I'll keep you posted if I get a chance to try it out this week.

**Comment by cjbottaro** [comment]
Yikes... so I started the using the branch myself today... and there were a lot of problems/errors.

If you `mix deps.update faktory_worker_ex` as of the time of this comment, it should be much more stable.

**Comment by emhagman** [comment]
We've also encountered this on Faktory Pro 0.9.5. Same stacktrace. We will try this out.

**Comment by cjbottaro** [comment]
Hi @emhagman. Yeah, the `v0.6.0` branch has been overhauled to be conceptually simpler, including a heartbeat process that has its own dedicated connection to the Faktory server.

I've been using that branch in staging, but not production... it seems to work fine. If you choose to try that branch, please note that configuration has been completely overhauled and simplified. It basically works exactly like Ecto repos now, but it's a breaking change from `v0.5.0`.

Side note, I've started the code for what will eventually be `v0.7.0` and in fact, `v0.6.0` might not even be released.

`v0.7.0` is basically a completely architectural overhaul with the goal of _massively simplified code_.

Each new version is bringing more simple and straightforward code, which in turn (I hope) will bring better reliability.

I'm hope to be using `v0.7.0` next week in staging.

**Comment by emhagman** [comment]
Sounds good 👍 I can wait for **v0.7.0**, I don't believe this is causing any actual issues.

**Comment by emhagman** [comment]
@cjbottaro Have you been using the v0.7.0 branch in staging? Any updates?

**Comment by cjbottaro** [comment]
@emhagman Yes, I have been using v0.7.0 in staging (and local dev), but that sees very little traffic.

The reason why I haven't released it yet is that I rewrote it all again using gen_stage (see the `gen_stage` branch). I suppose I should just make the gen_stage branch into v0.8.0 and get on with releasing v0.7.0.

**Comment by cjbottaro** [comment]
I need to review the code, tests, and documentation. I'll try to do that after work tonight and cut a release.

**Comment by cjbottaro** [comment]
@emhagman Can you try it out via Github? I made a tag for it, so you can add to your mix deps:
```
{:faktory_worker_ex, git: "https://github.com/cjbottaro/faktory_worker_ex", tag: "v0.7.0.0"}
```

**Comment by cjbottaro** [comment]
Ahh never mind, I published v0.7.0 to hex.

**Remember there are breaking changes** (regarding configuration).

**Comment by emhagman** [comment]
@cjbottaro Thanks, I will give this a shot. Appreciate the hard work 👍
